### PR TITLE
updating module S3 Bucket Keys for SSE-KMS

### PR DESCRIPTION
**Pull request by chirag1603**
##### SUMMARY

refrence: https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucket-key.html
Adding Parameter to enable to s3 bucket keys only when the encryption is aws:kms

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

s3 bucket keys

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: plugins/modules/s3_bucket.py [hunk 1]
```diff
@@ -84,6 +84,12 @@
     description: KMS master key ID to use for the default encryption. This parameter is allowed if I(encryption) is C(aws:kms). If
                  not specified then it will default to the AWS provided KMS key.
     type: str
+  encryption_bucket_key:
+    description: Bucket Key for SSE-KMS. This parameter is allowed only if encryption is aws:kms otherwise it will throw error.
```
Comment (tremble) [inline plugins/modules/s3_bucket.py:88]:
```suggestion
    description:
      - Enable S3 Bucket Key for SSE-KMS on new objects.
      - See the AWS documentation for more information
        U(https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucket-key.html).
      - Bucket Key encryption is only supported if I(encryption=aws:kms).
```

File: plugins/modules/s3_bucket.py [hunk 2]
```diff
@@ -725,6 +765,35 @@ def put_bucket_encryption(s3_client, bucket_name, encryption):
     server_side_encryption_configuration = {'Rules': [{'ApplyServerSideEncryptionByDefault': encryption}]}
     s3_client.put_bucket_encryption(Bucket=bucket_name, ServerSideEncryptionConfiguration=server_side_encryption_configuration)
 
+def put_bucket_key_with_retry(module, s3_client, name, expected_encryption):
+    max_retries = 3
+    for retries in range(1, max_retries + 1):
+        try:
```
Comment (tremble) [inline plugins/modules/s3_bucket.py:771]: It shouldn't be necessary to call put_bucket_encryption multiple times, I suspect the waiter just needs to be more patient, sometimes settings do take a couple of minutes to propagate.

**Review by tremble — CHANGES_REQUESTED** [review]
Thanks for taking the time to submit this PR.  I've added some initial comments.

Please take a look at the failing tests (sanity and integration).  I suspect the failed integration tests are because the default value was set to False, rather than setting no default.

Additionally, please rebase your PR, it looks like there's a lot of temporary debugging and merge commits: https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase

File: plugins/modules/s3_bucket.py (hunk 1 above)
Comment (tremble) [reply plugins/modules/s3_bucket.py:88]: Not for other reviewers - "S3 Bucket Keys for SSE-KMS" feels a little clumsy, however that's the name of the feature/product.

**Comment by chirag1603** [comment]
> @chirag1603 this PR contains the following merge commits:
> 
> * [1c8a770](https://github.com/ansible-collections/amazon.aws/commit/1c8a77064a4578352bf8e999a75ad5fa90fec2a8)
> * [92218d3](https://github.com/ansible-collections/amazon.aws/commit/92218d39b19c9f7807a64776b30b046109726b72)
> 
> Please [rebase your branch](http://docs.ansible.com/ansible/devel/dev_guide/developing_rebasing.html) to remove these commits.
> 
> [click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

Done

File: plugins/modules/s3_bucket.py [hunk 3]
```diff
@@ -84,6 +84,15 @@
...
     type: str
+  bucket_key_enabled:
+    description:
+      - Enable S3 Bucket Keys for SSE-KMS on new objects.
+      - See the AWS documentation for more information
+        U(https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucket-key.html).
+      - Bucket Key encryption is only supported if I(encryption=aws:kms).
+    required: false
+    type: bool
+
```
Comment (alinabuzachis) [inline plugins/modules/s3_bucket.py:95]: Please add `version_added: 5.0.0`

File: plugins/modules/s3_bucket.py [hunk 4]
```diff
@@ -214,6 +223,11 @@
     encryption: "aws:kms"
     encryption_key_id: "arn:aws:kms:us-east-1:1234/5678example"
 
+# Create a bucket with aws:kms encryption, Bucket key
+- amazon.aws.s3_bucket:
+    name: mys3bucket
+    bucket_key_enabled: true
```
Comment (tremble) [inline plugins/modules/s3_bucket.py:229]:
```suggestion
- amazon.aws.s3_bucket:
    name: mys3bucket
    bucket_key_enabled: true
    encryption: "aws:kms"
```

File: plugins/modules/s3_bucket.py (hunk 3 above)
Comment (tremble) [inline plugins/modules/s3_bucket.py:95]:
```suggestion
    type: bool
    version_added: 4.1.0
```

File: plugins/modules/s3_bucket.py [hunk 5]
```diff
@@ -284,6 +298,7 @@
     type: dict
     returned: I(state=present)
     sample: {
+
```
Comment (tremble) [inline plugins/modules/s3_bucket.py:301]:
```suggestion
```

**Review by tremble — CHANGES_REQUESTED** [review]
Looks like we're heading in the right direction...

There's a couple of minor docs issues, however it also looks like what's being returned in `encryption` isn't compatible with what's returned as for simple SSE

**Review by tremble — APPROVED** [review]

File: CHANGELOG.rst [hunk 6]
```diff
@@ -348,6 +348,7 @@ Minor Changes
 - s3_bucket - Tests for compatability with older versions of the AWS SDKs have been removed (https://github.com/ansible-collections/amazon.aws/pull/442).
 - s3_bucket - add new option ``object_ownership`` to configure object ownership (https://github.com/ansible-collections/amazon.aws/pull/311)
 - s3_bucket - updated to use HeadBucket instead of ListBucket when testing for bucket existence (https://github.com/ansible-collections/amazon.aws/pull/357).
+- s3_bucket - updated module to enable S3 Bucket Keys for SSE-KMS (https://github.com/ansible-collections/amazon.aws/pull/882).
```
Comment (alinabuzachis) [inline CHANGELOG.rst:351]: This file is autogenerated. There's no need to modify this one.

**Review by alinabuzachis — COMMENTED** [review]
@chirag1603 Thank you for taking time to work on this. Could you please add a changelog fragment https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to?

File: CHANGELOG.rst (hunk 6 above)
Comment (tremble) [reply CHANGELOG.rst:351]: I missed that.  @chirag1603 It should be a **fragment** file in changelog/fragments

**Pull request merged**